### ISSUE TITLE
[clang-tidy] Not possible to disable specific clang-analyzer checks

### ISSUE THREAD

**Issue by carlosgalvezp** [llvm/llvm-project#59588]
Running clang-tidy as:

```shell
clang-tidy main.cpp -checks=-*,clang-analyzer-core*,-clang-analyzer-core.uninitialized.Assign --list-checks
```

Displays:
```
    clang-analyzer-core.CallAndMessage
    clang-analyzer-core.CallAndMessageModeling
    clang-analyzer-core.DivideZero
    clang-analyzer-core.DynamicTypePropagation
    clang-analyzer-core.NonNullParamChecker
    clang-analyzer-core.NonnilStringConstants
    clang-analyzer-core.NullDereference
    clang-analyzer-core.StackAddrEscapeBase
    clang-analyzer-core.StackAddressEscape
    clang-analyzer-core.UndefinedBinaryOperatorResult
    clang-analyzer-core.VLASize
    clang-analyzer-core.builtin.BuiltinFunctions
    clang-analyzer-core.builtin.NoReturnFunctions
    clang-analyzer-core.uninitialized.ArraySubscript
    clang-analyzer-core.uninitialized.Assign                                   <<<<<<<<<< Should be removed
    clang-analyzer-core.uninitialized.Branch
    clang-analyzer-core.uninitialized.CapturedBlockVariable
    clang-analyzer-core.uninitialized.NewArraySize
    clang-analyzer-core.uninitialized.UndefReturn
```

This behavior does not happen on other checks.

**Comment by firewave** [comment]
They can be disabled as a whole via `-checks=-clang-analyzer-*` though.

**Comment by carlosgalvezp** [comment]
Yep. I do want to disable only specific checks that are problematic/have false positives, not all of them...

Could it be the dots in the name of the check which is the problem? The other checks don't have dots in the name.

**Comment by carlosgalvezp** [comment]
If I add `--explain-config`, I see something interesting:

```
'clang-analyzer-core.uninitialized.ArraySubscript' is enabled in the command-line option '-checks'.
'clang-analyzer-core.uninitialized.Assign' is enabled in the clang-tidy binary.
```

So the check is enabled in the binary, not via `-checks`, and can't be overriden.

**Comment by firewave** [comment]
That appears to be a certain type of static analyzer check - hence https://github.com/llvm/llvm-project/tree/main/clang/include/clang/StaticAnalyzer/Core/PathSensitive.

**Comment by carlosgalvezp** [comment]
@haoNoQ Do you have any insight as to why `core` CSA checks must be enabled for all other checks to work? Is the comment still valid nowadays?

```cpp
  // List all static analyzer checkers that our filter enables.
  //
  // Always add all core checkers if any other static analyzer check is enabled.
  // This is currently necessary, as other path sensitive checks rely on the
  // core checkers.
  for (StringRef CheckName : RegisteredCheckers) {
    std::string ClangTidyCheckName((AnalyzerCheckNamePrefix + CheckName).str());

    if (CheckName.startswith("core") ||
        Context.isCheckEnabled(ClangTidyCheckName)) {
      List.emplace_back(std::string(CheckName), true);
    }
  }
```

**Comment by carlosgalvezp** [comment]
This was first introduced around 10 years ago [here](https://github.com/llvm/llvm-project/commit/fe3f34e50b1c3e84728cac6e7ae06fd41402515f). I don't see many details in the commit message to tell what exactly is it that was failing in the past that needed this to be added.  @r4nt do you recall? Is this comment valid today?

**Comment by vbvictor** [comment]
From analyzer side, I recall some core checks could be moved to custom or visa versa.
But in general all core checks are needed AFAIK so nothing to fix (hence added Feature label)

**Comment by steakhal** [comment]
The clang static analyzer is designed in such a way that we expect its `core` checkers be enabled. The only reason they can be forcefully disabled when using the `clang` frontend is to unblock users in case something terribly goes wrong and the analyzer crashes otherwise. The problem is, the analyzer may also crash due to disabling a `core` checker.

Usually people don't disable `core` checkers when using the `clang` frontend (especially not by accident) but I think it's far far more likely that Clang Tidy users would accidentally blanket disable the `core` checkers and selectively enable just one checker expecting things to work - well, it won't.

So the predicament here is that I'd want users to have the ability to disable even core checkers (even when using the static analyzer via clang-tidy) but unless we have a user-story for doing this in a safe way. And I don't think I have a proposal for doing this safely.

---

If the user has problems with a core checker, file a separate issue against the clang static analyzer with that.

**Comment by NagyDonat** [comment]
> Digging deep into the code, I find:
> 
>   // Always add all core checkers if any other static analyzer check is enabled.
>   // This is currently necessary, as other path sensitive checks rely on the
>   // core checkers.
> 
> I don't quite understand the comment. What are `path sensitive checks`?

Most checkers in the static analyzer are "path sensitive checkers" that activate within the symbolic execution process performed by the static analyzer. The "symbolic execution" of a certain function `foo` is a process where:
- The statements in the body of `foo` are executed one by one
- In each elementary step the analyzer it updates a symbolic representation of the program state (e.g. "3" is written to the variable `x`, this function call returned a certain unknown ("symbolic") value, that expression evaluates to the sum of this particular symbol and the concrete integer "3" etc.)
- When it encounters a conditional statement, it "splits the state" and follows the two execution paths separately (this is what is highlighted in the name "path sensitive").
- It can enter ("inline") function calls if the definition is available (also executing their body step by step).
- Checkers can register various callbacks that are executed at certain points in this symbolic execution process, for example:
  - a `check::PreStmt<BinaryOperator>` callback is executed before each binary operator evaluation (e.g. can detect division by zero),
  - a `check::PostCall` callback is executed after each function call (e.g. can recognize library functions and update the state to record the effect of the call)
  - a `check::DeadDymbols` callback is executed when certain symbols go out of scope (e.g. can report when the result of `malloc` was not `free`d and can no longer be `free`d).

The analyzer will perform symbolic execution for each function that is not reached during the symbolic execution of another function. Note that **each function is executed symbolically only once and all checkers are active within that single execution**. This is advantageous because this way checkers can rely on state updates provided by other checkers.

The core checkers perform very basic checks (e.g. report division by zero or nullptr dereference), so disabling them is officially unsupported because other checkers might crash or behave unpredictably if their callbacks are activated in an erroneous state where e.g. a nullpointer was dereferenced but the symbolic execution still continues because the responsible checker was disabled.

My personal intuition says there are many core checkers that can be disabled without dangers, but there _may_ be some whose absence could trigger crashes. However, disabling core checkers is almost completely untested, so I cannot judge its effects accurately.

------

By the way it would be possible to tweak the behavior of the `core` checkers in a way that "disabling" them just suppresses their results without actually disabling them. That way users (e.g. tidy users) could easily disable any checker if they don't like its results, but they wouldn't be able to break anything by disabling `core` checkers (because the checker still runs in the background and silently discards execution paths where it would have generated an error).

**Comment by steakhal** [comment]
> By the way it would be possible to tweak the behavior of the `core` checkers in a way that "disabling" them just suppresses their results without actually disabling them. That way users (e.g. tidy users) could easily disable any checker if they don't like its results, but they wouldn't be able to break anything by disabling `core` checkers (because the checker still runs in the background and silently discards execution paths where it would have generated an error).

The fundamental problem here is that Tidy users rightfully expect modular and independent checks and the static analyzer doesn't work like that - at least not for the core checkers. With the rest of the checkers modularity is upheld.

This is why we had checker silencing - enabled by passing the `-analyzer-config silence-checkers="core.DivideZero;core.NullDereference"` options when using the static analyzer via the `clang` frontend. This will suppress the reports but keep the checker running behind the scenes.

One could transform the disablement action from tidy into a silencing option for the `core` checkers but forward those as `checker-disable=xxx` for any other kinds of checker. Probably this is the least evil.

**Comment by NagyDonat** [comment]
> One could transform the disablement action from tidy into a silencing option for the `core` checkers but forward those as `checker-disable=xxx` for any other kinds of checker. Probably this is the least evil.

Yes, that is also a valid approach, I could stand behind it.

However, I thought of an approach where "disabling a core checker just means silencing it" is universal and also applies in native execution of the analyzer. (For testing and debugging purposes we would probably still need a flag that really disables core checkers, but the users don't need to know about it.)

**Comment by vbvictor** [comment]
> This is why we had checker silencing - enabled by passing the -analyzer-config silence-checkers="core.DivideZero;core.NullDereference" options when using the static analyzer via the clang frontend. This will suppress the reports but keep the checker running behind the scenes.

So I suggest we keep this issue as tracker for that feature.

**Comment by steakhal** [comment]
> > One could transform the disablement action from tidy into a silencing option for the `core` checkers but forward those as `checker-disable=xxx` for any other kinds of checker. Probably this is the least evil.
> 
> Yes, that is also a valid approach, I could stand behind it.
> 
> However, I thought of an approach where "disabling a core checker just means silencing it" is universal and also applies in native execution of the analyzer. [...]

If we would uniformly silence other checkers, then there would be a weird behavior when we blanket enable a checker package, but "disable" each element of that checker package and still see the side-effects in the ExplodedGraph of all of those checkers.

For example, the `cplusplus.Move` has some interplay with the `cplusplus.InnerPointer` checker.
Let's say we would have two tidy invocations like:
- `clang-analyzer-unix.Malloc,clang-analyzer-cplusplus-*,-cplusplus.InnerPointer`
- `clang-analyzer-unix.Malloc,clang-analyzer-cplusplus-1,2,3,...,` (explicitly enabling each `cplusplus` checker except for the `InnerPointer` checker.

The user would expect the same outcome: no report differences; and of course no reports from the InnerPointer checker.
However, in practice the user could see different reports across the two configs (neither would have InnerPointer reports though).
If is because the fact that the InnerPointer checker is enabled can have (and at scale will have) side effects on the global Exploded Graph shared across the checkers. And this is probably not something the users would think of. Of course, they likely would never notice this subtlety.

The way around this was to actually forward the disable commands as `checker-disable` - which would truly disable the checker; turning the two configurations equivalent.
